crates/flox/src/commands/auth.rs

```diff
@@ -0,0 +1,295 @@
+    let thost = match host {
+        Some(string) => {
+            my_string = string;
+            Some(my_string.as_str())
+        },
+        None => None,
+    };
+
+    let mut flow = DeviceFlow::start(client_id.as_str(), thost).await?;
+
+    // eprintln!("res is {:?}", res);
+    eprintln!(
+        "Please visit {} in your browser",
+        flow.verification_uri.clone().unwrap()
+    );
+    eprintln!("And enter code: {}", flow.user_code.clone().unwrap());
+
+    thread::sleep(FIVE_SECONDS);
+
+    flow.poll(20).await
+}
+
+#[derive(Debug, Clone)]
+pub enum DeviceFlowState {
+    Pending,
+    Processing(time::Duration),
+    Success(Credential),
+    Failure(DeviceFlowError),
+}
+
+#[derive(Clone)]
+pub struct DeviceFlow {
+    pub host: String,
+    pub client_id: String,
+    pub user_code: Option<String>,
+    pub device_code: Option<String>,
+    pub verification_uri: Option<String>,
+    pub state: DeviceFlowState,
+}
+
+const FIVE_SECONDS: time::Duration = time::Duration::new(5, 0);
+
+impl DeviceFlow {
+    pub fn new(client_id: &str, maybe_host: Option<&str>) -> Self {
+        Self {
+            client_id: String::from(client_id),
+            host: match maybe_host {
+                Some(string) => String::from(string),
+                None => String::from("github.com"),
+            },
+            user_code: None,
+            device_code: None,
+            verification_uri: None,
+            state: DeviceFlowState::Pending,
+        }
+    }
+
+    pub async fn start(
+        client_id: &str,
+        maybe_host: Option<&str>,
+    ) -> Result<DeviceFlow, DeviceFlowError> {
+        let mut flow = DeviceFlow::new(client_id, maybe_host);
+
+        flow.setup().await;
+
+        match flow.state {
+            DeviceFlowState::Processing(_) => Ok(flow.to_owned()),
+            DeviceFlowState::Failure(err) => Err(err),
+            _ => Err(credential_error(
+                "Something truly unexpected happened".into(),
+            )),
+        }
+    }
+
+    pub async fn setup(&mut self) {
+        let body = format!("client_id={}", &self.client_id);
+        let entry_url = format!("https://{}/login/device/code", &self.host);
+
+        if let Some(res) = send_request(self, entry_url, body).await {
+            if res.contains_key("error") && res.contains_key("error_description") {
+                self.state = DeviceFlowState::Failure(credential_error(
+                    res["error_description"].as_str().unwrap().into(),
+                ))
+            } else if res.contains_key("error") {
+                self.state = DeviceFlowState::Failure(credential_error(format!(
+                    "Error response: {:?}",
+                    res["error"].as_str().unwrap()
+                )))
+            } else {
+                self.user_code = Some(String::from(res["user_code"].as_str().unwrap()));
+                self.device_code = Some(String::from(res["device_code"].as_str().unwrap()));
```

Comment:
non-blocking(error handling): If thw format is unexpected in any way this will panic due to `[".."]`